### DISCUSSION TITLE
How to get diagram editor closed notification/Action at server side

### DISCUSSION THREAD

**Discussion by sathya-1994** [eclipse-glsp/glsp#893] [General]
I am using https://github.com/eclipse-glsp/glsp-vscode-integration

I want notification/Action at java server side when we close the diagram editor (webview panel).

I tried to send DiagramEditorClosedAction using webviewPanel.onDidDispose() method but DiagramEditorClosedAction  was not sent to server side.

Code:
webviewPanel.onDidDispose(e => {
     const message = {
              clientId: clientId,
              action: new DiagramEditorClosedAction (),
               __localDispatch: true
     }
      webviewPanel.webview.postMessage(message);
}

**Comment by tortmayr** [comment]
Hi @sathya-1994,
once the diagram editor is closed you can no longer dispatch any actions to the server because the corresponding session on the server side is already disposed.

You could implement a [`ClientSessionListener`](https://github.com/eclipse-glsp/glsp-server/blob/6d5bb5ae30d8b0bdaff4d2ea2ec667df75be127e/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/session/ClientSessionListener.java#L46 )on the server side and react to the `sessionDisposed` event, which will always be triggered if the diagram editor has been closed on the client-side. 
(Note: The listener has to be registered via the [`GLSPServerManager.addListener`](https://github.com/eclipse-glsp/glsp-server/blob/6d5bb5ae30d8b0bdaff4d2ea2ec667df75be127e/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/session/ClientSessionManager.java#L77) method.

**Reply by sathya-1994** [reply]
@tortmayr ,

I used https://github.com/eclipse-glsp/glsp-server as server and https://github.com/eclipse-glsp/glsp-vscode-integration as client.

org.eclipse.glsp.server.features.sourcemodelwatcher.FileWatcher.sessionDisposed(ClientSession) method will be called when we close the vscode not when we close the editor.

**Reply by tortmayr** [reply]
The `sessionDiposed` method should definitely be called if the editor is closed. At first glance this sounds like a bug of the vscode-integration

**Reply by sathya-1994** [reply]
Also, I created a class by implementing ClientSessionListener interface. I added new listener using GLSPServerManager.addListener method (passed new class's object and client id as parameters).

sessionDisposed(ClientSession) of ClientSessionListener was not called when I closed editor at client side.

**Reply by tortmayr** [reply]
I can reproduce the issue and have created https://github.com/eclipse-glsp/glsp/issues/894 to track this
